internal/executor/job/processors/preempt_runs.go

```diff
@@ -0,0 +1,98 @@
+package processors
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+
+	executorContext "github.com/armadaproject/armada/internal/executor/context"
+	"github.com/armadaproject/armada/internal/executor/domain"
+	"github.com/armadaproject/armada/internal/executor/job"
+	"github.com/armadaproject/armada/internal/executor/reporter"
+	"github.com/armadaproject/armada/internal/executor/util"
+	"github.com/armadaproject/armada/pkg/api"
+)
+
+type RunPreemptedProcessor struct {
+	clusterContext   executorContext.ClusterContext
+	jobRunStateStore job.RunStateStore
+	eventReporter    reporter.EventReporter
+}
+
+func NewRunPreemptedProcessor(
+	clusterContext executorContext.ClusterContext,
+	jobRunStateStore job.RunStateStore,
+	eventReporter reporter.EventReporter,
+) *RunPreemptedProcessor {
+	return &RunPreemptedProcessor{
+		clusterContext:   clusterContext,
+		jobRunStateStore: jobRunStateStore,
+		eventReporter:    eventReporter,
+	}
+}
+
+func (j *RunPreemptedProcessor) Run() {
+	managedPods, err := j.clusterContext.GetBatchPods()
+	if err != nil {
+		log.Errorf("Failed to cancel runs because unable to get a current managed pods due to %s", err)
+		return
+	}
+
+	runsToCancel := j.jobRunStateStore.GetAllWithFilter(func(state *job.RunState) bool {
+		return state.PreemptionRequested
+	})
+	runPodInfos := createRunPodInfos(runsToCancel, managedPods)
+
+	util.ProcessItemsWithThreadPool(context.Background(), 20, runPodInfos,
+		func(runInfo *runPodInfo) {
+			pod := runInfo.Pod
+			if pod == nil {
+				// No pod to preempt
+				return
+			}
+
+			if util.IsInTerminalState(pod) {
+				// If pod is already finished, nothing to preempt
+				return
+			}
+
+			if !util.IsReportedPreempted(pod) {
+				err := j.reportPodPreempted(runInfo.Run, runInfo.Pod)
+				if err != nil {
+					log.Error(err)
```

Comment:
will this log something so we know the error has occurred as part of a preemption?

Reply:
Yes but I'll make it more explicit at this level